lib/browser.ts

```diff
@@ -624,116 +612,91 @@ export class ProtractorBrowser extends AbstractExtendedWebDriver {
    * @returns {!webdriver.promise.Promise} A promise that will resolve to the
    *    scripts return value.
    */
-  waitForAngular(opt_description?: string): wdpromise.Promise<any> {
+  async waitForAngular(opt_description?: string): Promise<any> {
     let description = opt_description ? ' - ' + opt_description : '';
     if (this.ignoreSynchronization) {
-      return this.driver.controlFlow().execute(() => {
-        return true;
-      }, 'Ignore Synchronization Protractor.waitForAngular()');
+      return true;
     }
 
-    let runWaitForAngularScript: () => wdpromise.Promise<any> = () => {
+    let runWaitForAngularScript = async(): Promise<any> => {
       if (this.plugins_.skipAngularStability() || this.bpClient) {
-        return this.driver.controlFlow().execute(() => {
-          return wdpromise.when(null);
-        }, 'bpClient or plugin stability override');
+        return null;
       } else {
-        // Need to wrap this so that we read rootEl in the control flow, not synchronously.
-        return this.angularAppRoot().then((rootEl: string) => {
-          return this.executeAsyncScript_(
-              clientSideScripts.waitForAngular, 'Protractor.waitForAngular()' + description,
-              rootEl);
-        });
+        let rootEl = await this.angularAppRoot();
+        return this.executeAsyncScript_(
+            clientSideScripts.waitForAngular, `Protractor.waitForAngular() ${description}`, rootEl);
       }
     };
 
-    return runWaitForAngularScript()
-        .then((browserErr: Function) => {
-          if (browserErr) {
-            throw new Error(
-                'Error while waiting for Protractor to ' +
-                'sync with the page: ' + JSON.stringify(browserErr));
+    let browserErr = await runWaitForAngularScript();
+    if (browserErr) {
+      throw new Error(
+          'Error while waiting for Protractor to ' +
+          'sync with the page: ' + JSON.stringify(browserErr));
+    }
+    await this.plugins_.waitForPromise(this);
+    try {
+      await this.driver.wait(async () => {
+        let results = await this.plugins_.waitForCondition(this);
+        return results.reduce((x, y) => x && y, true);
+      }, this.allScriptsTimeout, 'Plugins.waitForCondition()');
+    } catch (err) {
+      let timeout: RegExpExecArray;
+      if (/asynchronous script timeout/.test(err.message)) {
```

Comment:
Might as well extract this into a private function to make it more readable.

Reply:
Same with this.